### DISCUSSION TITLE
Feeding Overlay with Custom Metadata

### DISCUSSION THREAD

**Discussion by Timmy-485** [time4tea/gopro-dashboard-overlay#85] [Q&A]
I got the overlay up and working, but now I'm trying to feed custom data apart from that extracted from the GoPro camera to the widgets to be displayed.
I came across the use of the custom gpx file to feed said information, but the use of the gpx file seems limited. 
I have a CSV file with these fields [Timestamp, Speed, Alt, Lat, Long], which I then translate to a gpx file. 

The gpx format does not directly take speed values, apparently (the script seems to calculate the speeds from the provided timestamp and GPS locations)

What I want to know is: 
1. Is there a way to feed the overlay with specific Speed and altitude values? (my CSV values are more accurate than the GoPro-generated ones)

2. When a video segment is cut out from a video, both journey, and moving map widgets display data corresponding to the cutout portion. Can the journey map still be fed with the entire video's lat/lon meta-data while the moving map displays the cut metadata?

**Comment by time4tea** [comment]
Hiya,
Currently, you can't send data in as a csv file, although we have discussed it a bit previously.
The gpx format does have entries for altitude and speed though, if you can convert your csv file to gpx. The relevant tags are `ele` and `speed`

I haven't got much suggestion for the cropping right this second, I'll have a think.

Thanks for your interest!

**Reply by KyleGW** [reply]
A solution for this where the widgets support a start and end time might also enable a solution for Issue 44. 

display_start and display_end control when the map widget appears on the overlay. Useful for say some setup at beginning of the video where there's really no movement or when the GPS data is unreliable, you don't show a map there. Possibly even add some default minimum GPS precision option?

journey_data_start and journey_data_end control how much of the journey is drawn on the map. This will solve for Timm-485 request. 

````
<component type="moving_map" name="moving_map" x="1644" y="100" size="256" zoom="16" corner_radius="35" display_start="00:00:00" display_end="00:20:00"/><br />
<component type="moving_map" name="moving_map" x="1644" y="100" size="256" zoom="15" corner_radius="35" display_start="00:20:00" display_end="00:25:00"/>  
<component type="moving_map" name="moving_map" x="1644" y="100" size="256" zoom="13" corner_radius="35" display_start="00:25:00" />  
````
omitting display_end in last line above means it displays until end of the video

Then say I want to show a map for the first 20 minutes of the video that draws only that first 20 minutes of the GPS data, and not the full track. (This zooms journey map in a way) Then for the rest of the video (at 20 minute mark onwards) I want to show a map that shows the entire zoomed out view of the full track. I would use the layout below:

````
<component type="journey_map" name="journey_map" x="1644" y="376" size="256" corner_radius="35" display_start="00:00:00" display_end="00:20:00" journey_data_start="00:00:00" journey_data_end="00:20:00"/>
<component type="journey_map" name="journey_map" x="1644" y="376" size="256" corner_radius="35" display_start="00:20:00" display_end="01:00:00" journey_data_start="00:00:00" journey_data_end="01:00:00"/>
````

I don't know maybe tying this type of thing into the layout xml isn't the best idea, and there is some other cleaner way to accomplish the same thing?

**Reply by time4tea** [reply]
I too have been thinking along these lines, but I worry that making the xml too complex will be usable only by one or two people...

I dont want to reimplement spring in python to configure this software :-)

I really value your thoughts on this.. I'll see if I can find a way to make it do-able...

**Reply by Timmy-485** [reply]
On the first issue, I have added the speed tag to the gpx file:

```
         <trkpt lat ="47.37474684445815" lon ="8.551214607099784">
	    <ele>532.364117429042</ele>
	    <time>2022-05-26T17:29:51Z</time>
	    <speed>4.8</speed>
	  </trkpt>
	  <trkpt lat ="47.37474684445815" lon ="8.551214607099782">
	    <ele>532.589988881145</ele>
	    <time>2022-05-26T17:29:51Z</time>
	    <speed>4.95</speed>
	  </trkpt>
	  <trkpt lat ="47.37474684445815" lon ="8.551214607099778">
	    <ele>532.647805265802</ele>
	    <time>2022-05-26T17:29:51Z</time>
	    <speed>5.11</speed>
	  </trkpt>
	  <trkpt lat ="47.37474684445815" lon ="8.551214607099768">
	    <ele>532.581371583859</ele>
	    <time>2022-05-26T17:29:51Z</time>
	    <speed>5.29</speed>
	  </trkpt>
	  <trkpt lat ="47.37474684445815" lon ="8.551214607099766">
	    <ele>532.433775895698</ele>
	    <time>2022-05-26T17:29:51Z</time>
	    <speed>5.45</speed>
	  </trkpt>
	  <trkpt lat ="47.37474684445815" lon ="8.551214607099778">
	    <ele>532.505425198098</ele>
	    <time>2022-05-26T17:29:51Z</time>
	    <speed>5.64</speed>
	  </trkpt>
	  <trkpt lat ="47.37474684445815" lon ="8.551214607099785">
	    <ele>532.888202817488</ele>
	    <time>2022-05-26T17:29:51Z</time>
	    <speed>5.88</speed>
	  </trkpt>
	  <trkpt lat ="47.37474684445815" lon ="8.55121460709978">
	    <ele>532.346717074032</ele>
	    <time>2022-05-26T17:29:51Z</time>
	    <speed>6.08</speed>
	  </trkpt>
	  <trkpt lat ="47.37474684445815" lon ="8.551214607099768">
	    <ele>531.827193096863</ele>
	    <time>2022-05-26T17:29:51Z</time>
	    <speed>6.22</speed>
	  </trkpt>
	  <trkpt lat ="47.37474684445815" lon ="8.551214607099768">
	    <ele>531.918935263291</ele>
	    <time>2022-05-26T17:29:51Z</time>
	    <speed>6.21</speed>
```

Then I run the command: `python3 gopro-dashboard.py GOPR0186.mp4 output.mp4  --exclude thermometer temperature heartbeat cadence gradient --layout-xml /home/timmy/Desktop/Widgets/gopro-dashboard-overlay/gopro_overlay/layouts/default-1920x1080.xml --gpx /home/timmy/Desktop/Widgets/gopro-dashboard-overlay/bin/gpxfile.gpx --use-gpx-only --overlay-size 1920x1080`

But for some reason, the speed values I have supplied don't show up. It still uses the calculated speed from the GPS coordinates (if I'm not mistaken) and prints that to the widgets. Not sure what I missed

**Reply by time4tea** [reply]
Ah - my bad - I've been giving you bad information. While elevation is parsed as I said - speed was removed from GPX in version 1.1 and this code didn't parse it anyway. However, Garmin GPX devices can support speed as an extension field, and we already support some of those fields, so I could add support for parsing that...

It would look like this (from https://gis.stackexchange.com/questions/74222/what-are-the-units-for-garmins-gpx-gpxtpxspeed)
```
<gpx 
    xmlns="http://www.topografix.com/GPX/1/1" 
    xmlns:gpxtpx="http://www.garmin.com/xmlschemas/TrackPointExtension/v1" 
    version="1.1">
  <trk>
   <trkseg>
    <trkpt lat="38.900207" lon="-77.008990">
    <ele>11.67</ele>
    <time>2013-06-11T04:04:49Z</time>
    <extensions>
      <gpxtpx:TrackPointExtension>
        <gpxtpx:speed>6.86</gpxtpx:speed> 
      </gpxtpx:TrackPointExtension>
    </extensions>
    </trkpt>
    </trkseg>
  </trk>
</gpx>
```

This way we can keep the parsing as "standards-compliant" as possible.. (Updated to include namespace decls)
I'll do a new version now - it will take a few mins.

**Reply by time4tea** [reply]
There is some support for this in 0.74.0 - I didn't have a whole lot of time to test it, but it might work

**Reply by Timmy-485** [reply]
got it, I'll pull the new changes and try it out

**Reply by time4tea** [reply]
Just curious how you recorded speed/alt/location etc - what was generating CSV file? Most devices will output GPX or FIT - which we support - so what type of device is this?

**Reply by Timmy-485** [reply]
What i know is the GPMF data is extracted from GoPro video to generate those CSV's (not entirely certain about the whole process since I didn't handle the data extraction lol)

Also after the new changes I run into this while executing the script:
`Traceback (most recent call last):
  File "gopro-dashboard.py", line 98, in <module>
    config_dir = args.config_dir
AttributeError: 'Namespace' object has no attribute 'config_dir'`

**Reply by time4tea** [reply]
Oh dear. That will teach me for doing things a rush. Let me take a look. I don't understand why you are using a process that does GPMF -> CSV and then back into this program though. The gopro-dashboard program will extract the location information itself from the video file.  I'll take a look at the error right now.

**Reply by time4tea** [reply]
Hi @Timmy-485  - I'm not able to replicate your error.. How did you install the program? From your example command line, it doesn't look like you installed it with `pip`.
If I use:
```
python3 -mvenv venv
./venv/bin/pip install gopro-overlay
./venv/bin/gopro-dashboard.py GH010278.MP4 output.mp4  --gpx myfile.gpx --use-gpx-only --overlay-size 1920x1080
```
all is working... 
`Starting gopro-dashboard version 0.74.0` ...

**Reply by Timmy-485** [reply]
Oh, the CSV itself isn't generated for the overlay. The overlay implementation forms a small part of a larger project, lol. We use the CSV data to do other stuff along the pipeline, and since its reading seemed more accurate than the GoPro extracted data, I'm trying to feed that to the overlay itself.

I did install the earlier version with pip, not sure if I was to do the reinstallation after pulling the latest changes, but I'll just redo the installation to see

**Reply by time4tea** [reply]
OK.. so I guess you're using a version from git directly. That's fine - but unless you check out a tag you'll not always get stuff that works. It should, but YMMV. It would be good to know how you are calculating the speed, as then I could add it to this software and make it better for all. Do you have a clean git checkout?

**Reply by time4tea** [reply]
It would be great if you share a still of the generated overlay so I can put it in the gallery. Cheers!

**Reply by Timmy-485** [reply]
Ye, I was using the version from GitHub directly at first.
I installed it with pip now and got it to run, but the speed values are still not picked from the GPX file. 
Lon, Lat, and Alt all work perfectly, but speed still isn't picked from the GPX (it still seems to use the calculated one)

Sadly I didn't handle the metadata extraction, but from the info I've collected, it was a C script customized by one of the team members (which I can't share sadly. Private repo and all). What I know is it seems to stem from this repo [here](https://github.com/gopro/gpmf-parser )

Attached below are snippets of the GPX file as well as the overlay created 
![image](https://user-images.githubusercontent.com/82444135/211037266-4da9567b-2a9e-4bee-a289-6598978c0b04.png)

![image](https://user-images.githubusercontent.com/82444135/211036266-2fb1952b-ae08-4db9-a899-8175cc0085cf.png)

**Reply by time4tea** [reply]
OK, no worries. Maybe if you can send me  a gpx file that you generated, I can see what the issue might be... when I try a gpx file that *I* generate, I think it is working. But I might be wrong, or your file and mine might differ in a subtle way.
Of course your data will be kept private.
If that's OK, you can send it to me at gopro-overlay at time4tea.net 

Thanks for the image!!

**Reply by time4tea** [reply]
Hi @Timmy-485  - Thanks for the data - It showed a bug in the existing GPX code, but also in some other places - so very useful!
The issue is possibly fixed in v0.75.0  - please take a look.
btw - your conversion script is generating many values for the same second in the GPX file. GPX files usually have 1 second accuracy (due to ISO timestamp format - no fractions of a second), so most of the GPX file information will be discarded. The overlay program copes with it OK, it will just use the last point supplied with a duplicate time value.
You can also see that you teleport from one place to another- this is due probably to not considering DOP in GPS data..

**Reply by Timmy-485** [reply]
Yep, everything is working now, thank you.

That leaves the 2nd issue now lol.
"When a video segment is cut out from a video, both journey, and moving map widgets display data corresponding to the cutout portion. Can the journey map still be fed with the entire video's lat/lon meta-data while the moving map displays the cut metadata?"

The use case is as such:
I have a full GoPro video (which is mostly too long, so I make a summary of it basically)
Segments from the video are chopped out at certain intervals and then stitched back together to produce a "smaller" video.
But since each cut comes with its exclusive metadata, the journey map starts and ends at just the cut metadata, so after stitching the videos together, I have this disjoint journey map across the stitched video.

The other widgets, apart from the alt and journey map, work just fine with the cut metadata tho.

**Reply by time4tea** [reply]
You can't currently do this - I have thought about allowing some command-line parameters to control rendering just a section. However, a current work-around might be to just render the whole journey, and then cut out the sections you need from the rendered video. I realise this might be a bit slower, but it might be "good enough".

**Reply by Timmy-485** [reply]
Sadly, execution time is a tight constraint for me. Thus the workaround won't work for me.
What I was looking to do at first was try to adapt your codebase, find exactly where you're feeding each widget with the values, and see if I could take hold of that and supply the values as I wish, but I haven't made headway in that yet (if you can point me in the right direction that would be appreciated, if possible of course)


I also came across a little issue. Basically, I'm now trying to use an 854x480 video with the overlay
Run this command since the overlay-size argument can help 'force' the overlay to a size that isn't predefined in any of the XML files
`./venv/bin/gopro-dashboard.py files/GL011117.LRV files/output.mp4 --exclude thermometer temperature heartbeat cadence gradient --layout-xml files/default-1920x1080.xml --gpx files/gpxfile.gpx --use-gpx-only --overlay-size 854x480 
`

Overlay still comes out skewed tho

**Reply by time4tea** [reply]
I've never tried to use the LRV files! It should work, except that there won't be an overlay designed for that resolution.

You'll need to create one. 

When you use `overlay-size` it doesn't scale the position of the widgets, it just draws the overlay frame at the given size, but still the widgets will be in the wrong places

My suggestion would be use copy one of the default files, and use the gopro-overlay program to help edit it to move the widgets into the right places.

The gopro-overlay program watches the xml file and re-renders a single frame, to a file that you can open with a photo viewer. Photo viewers, on linux at least, reload the image when it changes.

**Reply by Timmy-485** [reply]
oh okay sure... I'll do that

**Reply by Timmy-485** [reply]
Hi, back again :)
Good news, I found a way to implement the issue of the journey map showing a complete journey map over cut metadata
Run into a bit of an issue now upon installing the program on a new PC, the version 0.79 i.e
```
timmy@AMALITECH-PC-00657:~$ venv/bin/gopro-dashboard.py /home/timmy/Desktop/Widgets/gopro/files/test_mauro/GOPR0186.mp4 gopro_gpmf.mp4
Starting gopro-dashboard version 0.79.0
    Timer(GPMD - Called: 1, Total: 0.08626, Avg: 0.08626, Rate: 11.59)
    Timer(extract GPS - Called: 1, Total: 1.26990, Avg: 1.26990, Rate: 0.79)
    Timer(extract ACCL - Called: 1, Total: 4.43663, Avg: 4.43663, Rate: 0.23)
    Timer(extract GRAV - Called: 1, Total: 0.02847, Avg: 0.02847, Rate: 35.13)
    Timer(extract CORI - Called: 1, Total: 0.02937, Avg: 0.02937, Rate: 34.05)
Timer(parsing - Called: 1, Total: 5.85096, Avg: 5.85096, Rate: 0.17)
Timer(loading timeseries - Called: 1, Total: 6.54409, Avg: 6.54409, Rate: 0.15)
Timer(program - Called: 1, Total: 6.54415, Avg: 6.54415, Rate: 0.15)
Traceback (most recent call last):
  File "venv/bin/gopro-dashboard.py", line 177, in <module>
    if counter.total() > 0:
AttributeError: 'ReasonCounter' object has no attribute 'total'
```

**Reply by time4tea** [reply]
Ah. I have inadvertently used a method from python 3.10.

I'll fix, as the target python is 3.8 but if you can use 3.10 it will work.

Sorry for that.

**Reply by Timmy-485** [reply]
No problem 
So quick question, can you point me to the function that handles the altitude chart widget 
Trying to take hold of it to make changes to the data being fed to that, but still trying to figure out which function handles that.

![image](https://user-images.githubusercontent.com/82444135/214590987-622600e9-8143-41c3-97d0-7dc5acac3057.png)
this widget, to be exact

**Reply by time4tea** [reply]
hi - are you able to share your changes? the  chart widget can graph any metric. see docs in https://github.com/time4tea/gopro-dashboard-overlay/tree/main/docs/xml/examples/07-chart
